cpp/src/arrow/dataset/partition.cc

```diff
@@ -562,6 +569,8 @@ inline Result<std::shared_ptr<Array>> CountsToOffsets(
 // since no Writers accept a selection vector.
 class StructDictionary {
  public:
+  static constexpr int32_t kMaxGroups = std::numeric_limits<int16_t>::max();
```

Comment:
Can you comment on this? It's not obvious why we're limited by the size of an `int16_t`.

Reply:
I picked it arbitrarily, to be honest. A huge number of groups seemed likely to be an error to see who would ask about it. Should we instead allow the full range of int32? @jorisvandenbossche

Reply:
For now I'll remove the constant `kMaxGroups` and allow any int32 group count.

Reply:
You never know if someone has a strange use case requiring a lot of groups, so if there is not a technical reason, I think it's good to just allow it

Reply:
I don't know, is a separate file created for each group? If so, it makes sense to put a configurable limit.

Reply:
Yes, at least one file for each group

Reply:
Then it's definitely worth having a reasonably small configurable limit (such as 100). I suspect it's easy to end up with Arrow creating a million files if you do a mistake in choosing your partition columns.

Reply:
As long as it is configurable, then that is fine for me. 
But I think something like 100 is too small. For example, the NYC taxi dataset partitioned by year + month for 8 years of data already has 8*12 = 96 groups. And I think partitioning by day is not that uncommon in practice for big data (although for those cases you will probably not write that all at once)

Reply:
I'll add max groups as a member of WriteOptions